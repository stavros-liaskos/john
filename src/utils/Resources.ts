import { components } from '../types/schema';

interface IResources {
  cacheMap: Record<string, unknown>;
  fetch: (url: string) => void;
}

class Resources implements IResources {
  private cacheMap;

  constructor() {
    this.cacheMap = new Map();
  }

  fetch(url) {
    if (this.cacheMap.has(url)) {
      return this.cacheMap.get(url);
    }

    const promise = fetch(url, {
      method: 'GET',
      credentials: 'include',
    })
      .then(response => {
        return response.json();
      })
      .then((data: components['schemas']['FollowedArtistsResponse']) => {
        this.cacheMap.set(url, data);
        return data;
      });

    throw promise;
  }
}

const resources = new Resources();

export default resources;
