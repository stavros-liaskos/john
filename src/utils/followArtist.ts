import { components } from '../types/schema';
import Endpoints from '../types/endpoints';

async function followArtist(artist: components['schemas']['SearchResultArtistDto'], cb: () => void) {
  const headers = new Headers({ 'Content-Type': 'application/json' });
  await fetch(Endpoints.Follow, {
    method: 'POST',
    credentials: 'include',
    headers,
    body: JSON.stringify(artist),
  })
    .then(() => {
      console.log(`${artist.name} followed successfully`);
    })
    .catch(error => {
      console.error('Error:', JSON.stringify(error));
    })
    .finally(() => {
      cb();
    });
}
export default followArtist;
