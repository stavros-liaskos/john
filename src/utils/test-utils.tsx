import React from 'react';
import { render, act } from '@testing-library/react';
import ThemeProvider from '../contexts/Theme/ThemeProvider';
import { UserProvider } from '@auth0/nextjs-auth0/client';
import ArtistsListProvider from '../contexts/ArtistsList/ArtistsListProvider';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemeProvider>
      <UserProvider>
        <ArtistsListProvider>{children}</ArtistsListProvider>
      </UserProvider>
    </ThemeProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: { [key: string]: unknown }) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };

export const renderWithAct = async (ui: React.ReactElement, options?: { [key: string]: unknown }) => {
  const component = render(ui, { wrapper: AllTheProviders, ...options });

  const promise = Promise.resolve();
  await act(() => promise);

  return component;
};

export const beforeEachTest = () => {
  jest.resetModules();
  jest.clearAllMocks();
  jest.restoreAllMocks();
  localStorage.clear();
};
