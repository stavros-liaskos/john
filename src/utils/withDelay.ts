export default function withDelay(cb: () => void, delay = 1000) {
  return setTimeout(() => {
    cb();
  }, delay);
}
