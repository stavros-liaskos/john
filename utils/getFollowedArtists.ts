import { components } from '../types/schema';
import { Dispatch, SetStateAction } from 'react';

export function getFollowedArtists(
  setState: Dispatch<SetStateAction<components['schemas']['FollowedArtistsResponse'] | undefined>>,
  setLoading: Dispatch<SetStateAction<boolean>>,
) {
  setLoading(true);
  const fetchData = async () => {
    const data = await fetch(`${process.env.BE_BASE_URL}/me/followed-artists`, {
      method: 'GET',
      mode: 'cors',
      headers: {
        'Content-Type': 'application/json',
      },
      referrerPolicy: 'no-referrer',
    });
    const json = await data.json();
    setState(json);
    setLoading(false);
  };
  fetchData().catch(console.error);
}
