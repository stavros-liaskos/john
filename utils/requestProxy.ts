import type { NextApiRequest, NextApiResponse } from 'next';
import headers from '../pages/api/headers';

export default async function requestProxy(req: NextApiRequest, res: NextApiResponse<unknown>) {
  const requestOptions = {
    method: req.method,
    headers,
    ...(req.body && { body: JSON.stringify(req.body) }),
  };

  try {
    const response = await fetch(process.env.BE_BASE_URL as unknown as URL, requestOptions);
    const result = await response.json();
    console.log(result);
    res.status(200).json(JSON.parse(result));
  } catch (e) {
    console.error('error', e);
  }
}
