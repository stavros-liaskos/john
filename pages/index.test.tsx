import React from 'react';
import Home from './index.page';
import { renderWithAct } from '../utils/test-utils';

describe('Home', () => {
  it('renders a heading', async () => {
    const component = await renderWithAct(<Home />);
    const { container } = component;

    // expect(component.getByText('Loading...')).toBe('Loading...');
    expect(container).toMatchSnapshot(); // TODO fix
  });
});
