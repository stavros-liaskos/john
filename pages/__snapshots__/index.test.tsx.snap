// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders a heading 1`] = `
<div>
  <div
    class="container max-auto"
  >
    <div
      class="sm:border-x-2 rr-border flex flex-col h-screen"
    >
      <header
        class="flex justify-between basis-14 rr-border border-b-2"
      >
        <button
          class="m-3"
          data-testid="moon"
        >
          <svg
            class="h-6 w-6"
            color="var(--geist-foreground)"
            fill="none"
            height="28"
            shape-rendering="geometricPrecision"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            width="28"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 12.79A9 9 0 1111.21 3 7 7 0 0021 12.79z"
            />
          </svg>
        </button>
      </header>
      <main
        class="flex-1 flex flex-col items-center w-full p-3 h-24"
      >
        <div
          class="flex flex-col items-center w-full lg:w-9/12"
        >
          <div
            class="flex flex-col items-center justify-center lg:w-1/2"
          >
            <h2
              class="rr-text text-4xl mb-3"
            >
              Welcome to Release Raccoon!
            </h2>
            <p
              class="rr-text text-xl mb-3"
            >
              Receive your favorite artists' music in your email every week!
            </p>
            <a
              href="/api/auth/login"
            >
              <button
                class="btn btn-large lg:ml-8 my-8 w-44 h-14"
              >
                Register
              </button>
            </a>
          </div>
        </div>
      </main>
      <footer
        class="flex items-center justify-center basis-16 border-t-2 rr-border"
      >
        <a
          class="flex items-center rr-text"
          href="https://github.com/jaivalis/release-raccoon"
          rel="noopener noreferrer"
          target="_blank"
        >
          Powered by  
          <div
            class="mx-2"
          >
            <img
              alt="Github Logo"
              data-nimg="1"
              decoding="async"
              height="32"
              loading="lazy"
              src="/_next/image?url=%2FGitHub-Mark-32px.png&w=64&q=75"
              srcset="/_next/image?url=%2FGitHub-Mark-32px.png&w=32&q=75 1x, /_next/image?url=%2FGitHub-Mark-32px.png&w=64&q=75 2x"
              style="color: transparent;"
              width="32"
            />
          </div>
        </a>
      </footer>
    </div>
  </div>
</div>
`;
